src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -2335,20 +2340,79 @@ public sealed override void Filter(
                     IValueSet? whenFalseValues,
                     out Tests whenTrue,
                     out Tests whenFalse,
-                    ref bool foundExplicitNullTest)
+                    ref bool foundExplicitNullTest,
+                    BindingDiagnosticBag diagnostics)
                 {
                     var trueBuilder = ArrayBuilder<Tests>.GetInstance(RemainingTests.Length);
                     var falseBuilder = ArrayBuilder<Tests>.GetInstance(RemainingTests.Length);
-                    foreach (var other in RemainingTests)
+                    for (int i = 0; i < RemainingTests.Length; i++)
                     {
-                        other.Filter(builder, test, state, whenTrueValues, whenFalseValues, out Tests oneTrue, out Tests oneFalse, ref foundExplicitNullTest);
+                        var other = RemainingTests[i];
+                        other.Filter(builder, test, state, whenTrueValues, whenFalseValues, out Tests oneTrue, out Tests oneFalse, ref foundExplicitNullTest, diagnostics);
                         trueBuilder.Add(oneTrue);
                         falseBuilder.Add(oneFalse);
+
+                        if (i > 0
+                            && oneTrue is False
+                            && this is OrSequence { RemainingTests: [Not { Negated: var negated }, ..] }
+                            && isRecognizedPartOfNegated(test, negated))
```

Comment:
:memo: Surprisingly, removing this `isRecognizedPartOfNegated` check doesn't seem to impact any tests or bootstrapping.

Reply:
It feels like if we can't come up with any tests which depend on this filtering (e.g. scenarios where an unwanted warning is reported in absence of this check), then it bears further investigation. We should try to understand this area well enough to either identify such a scenario and test it, or to be confident that this check is not needed and remove it.

Reply:
I don't think it's possible. We would need a test that appears inside a condition inside `Tests.Not` but is not a requirement for that condition to succeed.
For example, `Or(Not(Not(Test)), RedundantTest` or `Or(Not(Or(Test, OtherTest)), RedundantTest)`. 
But neither of these structures are possible, as `not not Test` gets represented as `Test` and `not (Test or OtherTest)` gets represented as `And(Not(Test), Not(OtherTest))`.

So all the tests I've added to show the impact of this filtering logic only show that it limits useful warnings. I was not able to show the impact on false warnings.
Still, I feel safer keeping this filtering logic, in case I'm missing something or the above situation changed.

Reply:
Perhaps we could, without changing the behavior in this PR, add an assertion if the preceding logic holds, the `isRecognizedPartOfNegated` must also hold. So, if we ever find a situation where the implication is not met, we will at least know about it.